drivers/blk/sdmmc/src/main.rs

```diff
@@ -0,0 +1,279 @@
+#![no_std] // Don't link the standard library
+#![no_main] // Don't use the default entry point
+
+extern crate alloc;
+
+mod sddf_blk;
+
+use core::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll, RawWaker, RawWakerVTable, Waker},
+};
+
+use alloc::boxed::Box;
+use sddf_blk::{
+    blk_dequeue_req_helper, blk_enqueue_resp_helper, blk_queue_empty_req_helper,
+    blk_queue_full_resp_helper, blk_queue_init_helper, BlkOp, BlkRequest, BlkStatus,
+};
+use sdmmc_hal::meson_gx_mmc::MesonSdmmcRegisters;
+
+use sdmmc_protocol::sdmmc::{InterruptType, SdmmcHalError, SdmmcHardware, SdmmcProtocol};
+use sel4_microkit::{debug_print, debug_println, protection_domain, Channel, Handler, Infallible};
+
+const BLK_VIRTUALIZER: sel4_microkit::Channel = sel4_microkit::Channel::new(0);
+
+const INTERRUPT: sel4_microkit::Channel = sel4_microkit::Channel::new(1);
+
+const SDCARD_SECTOR_SIZE: u32 = 512;
+const SDDF_TRANSFER_SIZE: u32 = 4096;
+const SDDF_TO_REAL_SECTOR: u32 = SDDF_TRANSFER_SIZE / SDCARD_SECTOR_SIZE;
+
+const RETRY_CHANCE: u16 = 5;
+
+// Debug function for printing out content in one block
+#[allow(dead_code)]
+fn print_one_block(ptr: *const u8) {
+    unsafe {
+        // Iterate over the 512 bytes and print each one in hexadecimal format
+        for i in 0..512 {
+            let byte = *ptr.add(i);
+            if i % 16 == 0 {
+                debug_print!("\n{:04x}: ", i);
+            }
+            debug_print!("{:02x} ", byte);
+        }
+        debug_println!();
+    }
+}
+
+// No-op waker implementations, they do nothing.
+unsafe fn noop(_data: *const ()) {}
+unsafe fn noop_clone(_data: *const ()) -> RawWaker {
+    RawWaker::new(_data, &VTABLE)
+}
+
+// A VTable that points to the no-op functions
+static VTABLE: RawWakerVTable = RawWakerVTable::new(noop_clone, noop, noop, noop);
+
+// Function to create a dummy Waker
+fn create_dummy_waker() -> Waker {
+    let raw_waker = RawWaker::new(core::ptr::null(), &VTABLE);
+    unsafe { Waker::from_raw(raw_waker) }
+}
+
+#[protection_domain(heap_size = 0x10000)]
+fn init() -> HandlerImpl<'static, MesonSdmmcRegisters> {
```

Comment:
That is what I have said in the note, as I cannot control the voltage. I will not be able to init the card by myself. If someone boots the system from an eMMC but want to use this driver to use sdcard, it will not work because I do not have any way to control the voltage to boot the sdcard.

Reply:
The init of the card is easy to implement, providing I can control the voltage in the hardware layer, which I cannot do right now. For reset, it is also easy to implement but unnecessary because the card is already in sdcard high speed mode, reset does not have any benefits as I know.

Reply:
> I'm not talking about the voltages et al, I'm talking about the command sending flow.
> 
> There's a whole initialisation sequence for sending commands? Refer to my driver
> 
> Without it you can't initialise the SD layer on secondary MMC devices, even if the voltage is turned on (which at least for the imx8 it is).
> 
> (Also, hopefully this should support hotplug, where that's definitely necessary)

No need to worry about hotplugging right now. I am pretty sure it would be supported in the future. If people really want to have reset even when doing it makes no obvious benefits, I can implement that as well.

Reply:
> reset even when doing it makes no obvious benefits

We should not be relying on U-Boot to reset the device properly. We should be doing device reset ourselves.

Reply:
Sure, but let me provide a bit more context about this. The card setup function is always on my plan as to improve the read/write speed to above 100MB/s, a card setup would be definitely be needed. I was planning to add the card setup at the second iteration, when we have some way to change voltage. But since we do not want to completely rely on uboot set up the device for us, I can do it ahead. However, one thing to understand is that, whether I implement it or not we are all relying on uboot to set up the voltage, so we are still depending on uboot to set up the device, more or less.